contributors/design-proposals/storage/container-storage-interface-inline-volumes.md

```diff
@@ -0,0 +1,211 @@
+	// VolumeHandle is prefixed by UID of the namespace where the pod is located.
+	CSIVolumeHandlePrefixNamespace CSIVolumeHandlePrefix  = "Namespace"
+	// VolumeHandle is not modified.
+	CSIVolumeHandlePrefixNone CSIVolumeHandlePrefix = "None"
+)
+```
+
+The difference between `CSIVolumeSource` (in-lined in a pod) and `CSIPersistentVolumeSource` (in PV) are:
+
+* All secret references in in-line volumes can refer only to secrets in the same namespace where the corresponding pod is running. This is common in all other volume sources that refer to secrets, incl. Flex.
+* VolumeHandle in in-line volumes can have a prefix. This prefix (Pod UID, Namespace UID or nothing) is added to the VolumeHandle before each CSI call. It makes sure that each pod uses a different volume ID for its ephemeral volumes.  The prefix must be explicitly set by pod author, there is no default.
+	* Users don't need to think about VolumeHandles used in other pods in their namespace, as each pod will get an unique prefix when `CSIVolumeHandlePrefixPod` is used. CSI volume ID with this prefix cannot accidentally conflict by another volume ID in another pod.
+	* Each pod created by ReplicaSet, StatefulSet or DaemonSet will get the same copy of a pod template. `CSIVolumeHandlePrefixPod` makes sure that each pod gets its own unique volume ID and thus can get its own volume instance.
+	* Without the prefix, user could guess volume ID of a secret-like CSI volume of another user and craft a pod with in-line volume referencing it. CSI driver, obeying idempotency, must then give the same volume to this pod. If users can use only`CSIVolumeHandlePrefixNamespace` or `CSIVolumeHandlePrefixPod`in their in-line volumes, we can make sure that they can't steal secrets of each other.
+		* `PodSecurityPolicy` will be extended to allow / deny users using in-line volumes with no prefix.
+	* Finally, `CSIVolumeHandlePrefixNone` allows selected users (based on PSP) to use persistent storage volumes in-line in pods.
+
+## Implementation
+#### Provisioning/Deletion
+N/A, it works only with PVs and not with in-line volumes.
+
+### Attach/Detach
+Current `storage.VolumeAttachment` object contains only reference to PV that's being attached. It must be extended with VolumeSource for in-line volumes in pods.
+
+```go
+// VolumeAttachmentSpec is the specification of a VolumeAttachment request.
+type VolumeAttachmentSpec struct {
+    // <snip>
+
+	// Source represents the volume that should be attached.
+	Source VolumeAttachmentSource
+}
+
+// VolumeAttachmentSource represents a volume that should be attached, either
+// PersistentVolume or a volume in-lined in a Pod.
+// Exactly one member can be set.
+type VolumeAttachmentSource struct {
+	// Name of the persistent volume to attach.
+	// +optional
+	PersistentVolumeName *string
+
+	// InlineVolumeSource represents the source location of a in-line volume in a pod to attach.
+	// +optional
+    InlineVolumeSource *InlineVolumeSource
```

Comment:
What does inline mean here? Is this actually inline if the VolumeSource spec isn't embedded in the PodSpec?

Reply:
Pod yaml has an in-line volume:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: testpod
spec:
  containers:
    ...
  volumes:
      - name: vol
        csi:
          driver: io.kubernetes.storage.mock
          volumeAttributes:
              name: "Mock Volume 1"
          volumeHandle: "1"
```

This `csi:` gets copied to `VolumeAttachment` because Pod does not need to exist at detach time and we need the volume source somewhere to detach volumes.

Reply:
@jsafrane couple of things I would like to suggest here:
1) I don't like the name `InlineVolumeSource` as it seems to prematurely leak the usage/impl of component in its name.  Maybe 
2) I am thinking of removing one level of object abstraction by putting `PersistentVolumeName` and `Source` directly in the `VolumeAttachmentSpec`

```go
type VolumeAttachmentSpec struct {
    // <snip>
    // Source represents the volume that should be attached.
    PersistentVolumeName *string

    Source *VolumeAttachmentSource
}

type VolumeAttachmentSource struct {
	// VolumeSource is copied from the pod. It ensures that attacher has enough
	// information to detach a volume when the pod is deleted before detaching.
	// Only CSIVolumeSource can be set.
	// Required.
	VolumeSource v1.VolumeSource
 	// Namespace of the pod with in-line volume. It is used to resolve
	// references to Secrets in VolumeSource.
	// Required.
	Namespace string
}
```

Let me know if those changes work for you... If so, I can continue with implementation.